src/auth/shibboleth.js

```diff
@@ -14,7 +14,7 @@ module.exports = safeAsync(async (req, res) => {
   const user = await createOrUpdateUser(req, netid)
   addJwtCookie(req, res, user)
 
-  // Finally, we'll redirect the user to the queue homepage. They'll now
-  // have the required token to make authenticated requests.
-  res.redirect(withBaseUrl('/'))
+  // Finally, we'll redirect the user to either their original url or the queue homepage.
+  // They'll now have the required token to make authenticated requests.
+  res.redirect(req.query.redirect || withBaseUrl('/'))
```

Comment:
I think the redirect needs `withBaseUrl` as well - e.g. currently gets redirected to https://edudev.cs.illinois.edu/course/1 when it should be https://edudev.cs.illinois.edu/q/course/1. I think the same holds true on the client-side redirect for dev mode, although we shouldn't be serving dev mode from a non-root path at present. You can test locally with `BASE_URL=/q npm run dev` and then go to http://localhost:3000/q